contracts/button/button.rs

```diff
@@ -0,0 +1,399 @@
+            .returns::<Balance>()
+            .fire()?;
+        Ok(balance)
+    }
+
+    fn transfer_tx(
+        &self,
+        transfer_to_selector: [u8; 4],
+        to: AccountId,
+        value: u128,
+    ) -> core::result::Result<(), InkEnvError> {
+        build_call::<DefaultEnvironment>()
+            .call_type(Call::new().callee(self.get().button_token))
+            .exec_input(
+                ExecutionInput::new(Selector::new(transfer_to_selector))
+                    .push_arg(to)
+                    .push_arg(value),
+            )
+            .returns::<()>()
+            .fire()
+    }
+
+    fn check_role(&self, account: AccountId, role: Role) -> Result<()>
+    where
+        Self: AccessControlled,
+    {
+        <Self as AccessControlled>::check_role(
+            self.get().access_control,
+            account,
+            role,
+            |why: InkEnvError| {
+                Error::ContractCall(format!("Calling access control has failed: {:?}", why))
+            },
+            || Error::MissingRole,
+        )
+    }
+
+    fn allow(&mut self, player: AccountId, caller: AccountId, this: AccountId) -> Result<()>
+    where
+        Self: AccessControlled,
+    {
+        let required_role = Role::Admin(this);
+        self.check_role(caller, required_role)?;
+        self.get_mut().can_play.insert(player, &true);
+        Ok(())
+    }
+
+    fn bulk_allow(
+        &mut self,
+        players: Vec<AccountId>,
+        caller: AccountId,
+        this: AccountId,
+    ) -> Result<()>
+    where
+        Self: AccessControlled,
+    {
+        let required_role = Role::Admin(this);
+        self.check_role(caller, required_role)?;
+
+        for player in players {
+            self.get_mut().can_play.insert(player, &true);
+        }
+        Ok(())
+    }
+
+    fn disallow(&mut self, player: AccountId, caller: AccountId, this: AccountId) -> Result<()>
+    where
+        Self: AccessControlled,
+    {
+        let required_role = Role::Admin(this);
+        self.check_role(caller, required_role)?;
+
+        self.get_mut().can_play.insert(player, &false);
+        Ok(())
+    }
+
+    fn press(&mut self, now: u32, caller: AccountId) -> Result<()> {
+        let ButtonData {
+            can_play, presses, ..
+        } = self.get();
+
+        if self.is_dead(now) {
+            return Err(Error::AfterDeadline);
+        }
+
+        if presses.get(&caller).is_some() {
+            return Err(Error::AlreadyParticipated);
+        }
+
+        if !can_play.get(&caller).unwrap_or(false) {
+            return Err(Error::NotWhitelisted);
+        }
+
+        let score = self.score(now);
+
+        self.get_mut().presses.insert(&caller, &score);
+        self.get_mut().press_accounts.push(caller);
+        self.get_mut().last_presser = Some(caller);
+        self.get_mut().last_press = now;
+        self.get_mut().total_scores += score;
```

Comment:
is there any way of making a single read and a single write?

Reply:
that *should*^TM do it

Reply:
do you mean that this is compiled into single-read-single-write?

Reply:
I meant this: https://github.com/Cardinal-Cryptography/aleph-node/blob/blue-button/contracts/button/button.rs#L297
pulls the data from storage, mutates in-memory and writes again